[4365] Add Video Thumbnails Docs (#4373)
GetStream/stream-chat-android: docusaurus/docs/Android/04-compose/07-guides/02-customizing-image-and-video-previews.mdx

# Customizing Image and Video Previews

## Introduction

The Compose UI Components chat SDK supports various types of attachments by using [attachment factories](../03-general-customization/01-chat-theme.mdx#streamattachmentfactories). The one used to display image and video attachments is called `MediaAttachmentFactory` and offers a degree of customization.

:::note

> Video Attachment preview thumbnails are a paid feature and as such can be turned off. You can find the pricing [here](https://getstream.io/chat/pricing/) and the information on how to turn them off in the [this](../05-message-components/03-message-list.mdx#image-and-video) section.

Reviewer: Again, these are technical docs, do we need to mention pricing here?

Author: Same as here:

https://github.com/GetStream/stream-chat-android/pull/4373#discussion_r1029655108

            DefaultItemOverlayContent()
        }
    },
    previewItemOverlayContent: @Composable (attachmentType: String?) -> Unit = { attachmentType ->
        if (attachmentType == AttachmentType.VIDEO) {
            DefaultPreviewItemOverlayContent()
        }
    },
): AttachmentFactory =
    AttachmentFactory(
        canHandle = {
            it.none { attachment ->
                !attachment.isImage() && !attachment.isVideo()
            }
        },
        previewContent = { modifier, attachments, onAttachmentRemoved ->
            MediaAttachmentPreviewContent(
                attachments = attachments,
                onAttachmentRemoved = onAttachmentRemoved,
                modifier = modifier,
                previewItemOverlayContent = previewItemOverlayContent
            )
        },
        content = @Composable { modifier, state ->
            MediaAttachmentContent(
                modifier = modifier,
                attachmentState = state,
                maximumNumberOfPreviewedItems = maximumNumberOfPreviewedItems,
                itemOverlayContent = itemOverlayContent
            )
        }
    )
```

> * `maximumNumberOfPreviewedItems`: Controls the maximum amount of tiles that can be displayed when previewing attachments inside the message list. In case this value is smaller than the number of attachments contained in the message, a count informing us how many more are left is displayed.

Reviewer: ```suggestion
* `maximumNumberOfPreviewedItems`: Controls the maximum amount of tiles that can be displayed when previewing attachments inside the message list. In case there are more attachments inside the message then we are previewing, a count of the remaining items is displayed.
```

Author: Used the given example and slightly built on top of it:

```
* `maximumNumberOfPreviewedItems`: Controls the maximum number of tiles that can be displayed when previewing attachments inside the message list. In case there are more attachments inside the message than are able to be previewed, the remaining item count is displayed.
```

        // Apply it only to video attachments
        if (attachmentType == AttachmentType.VIDEO) {
            CustomPlayButton(
                modifier = Modifier
                    .padding(2.dp)
                    .background(
                        color = Color(red = 256, blue = 256, green = 256, alpha = 220),
                        shape = RoundedCornerShape(8.dp)
                    )
                    .fillMaxWidth(0.35f)
                    .aspectRatio(1.20f),
            )
        }
    })
```

Since we only want to modify `MediaAttachmentFactory` and keep the rest of the default factories, we'll copy them in the same order they are created in [StreamAttachmentFactories.defaultFactories()](https://github.com/GetStream/stream-chat-android/blob/main/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/StreamAttachmentFactories.kt#L65).

```kotlin
val attachmentFactories = listOf(
    UploadAttachmentFactory(),
    LinkAttachmentFactory(linkDescriptionMaxLines = 5),
    GiphyAttachmentFactory(),
    customMediaAttachmentFactory,
    FileAttachmentFactory(),
    UnsupportedAttachmentFactory()
)
```

> Remember, order of precedence matters when creating a list of attachment factories, as the first one whose `canHandle()` function evaluates to true will render the given attachment.

Reviewer: ```suggestion
Remember, ordering matters when creating a list of attachment factories. The first factory that returns true on `canHandle()` will be used to render the given attachment.
```

Too verbose, can be simplified.

Author: The original one is only 1 word and 11 characters longer than the suggestion, so I'll keep it as it's grammatically more correct.

Reviewer: Even though it is grammatically correct i find it too convoluted and hard to read.

Finally, let's replace the default attachment factories:

```kotlin
override fun onCreate(savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
    super.onCreate(savedInstanceState, persistentState)

    setContent {
        // Replace the default attachment factories
        ChatTheme(attachmentFactories = attachmentFactories) {
            MessagesScreen(
                viewModelFactory = messageListViewModelFactory,
                onBackPressed = { finish() }
            )
        }
    }
}
```

And the resulting experience looks like this:

![Custom Image and Video Attachment Previews](../../assets/compose_customizing_image_and_video_attachments_custom.png)

## Conclusion

> If you need to put small personal touches to how your app displays image and video attachments while expending very little time on development, the approach we've discussed is optimal. If you need to do large changes, remember you can always create your own attachment factory and replace the default one with it.

Reviewer: There are guides how to create a custom attachment. Maybe link it somewhere?

Author: Added a note below.
